ring-core/src/ring/middleware/file.clj

```diff
@@ -40,15 +40,29 @@
   Accepts the following options:
 
   :index-files?    - look for index.* files in directories, defaults to true
-  :allow-symlinks? - serve files through symbolic links, defaults to false"
+  :allow-symlinks? - serve files through symbolic links, defaults to false
+  :prefer-handler? - give preference to handler to process the file. 
+                     if the handler successfully processes the file, do not serve
+                     static content"
   ([handler root-path]
    (wrap-file handler root-path {}))
   ([handler root-path options]
    (ensure-dir root-path)
    (fn
      ([request]
-      (or (file-request request root-path options) (handler request)))
+      (if (:prefer-handler? options)
+        (let [response (handler request)]
+          (if (not= 404 (:status response))
+            response
+            (file-request request root-path options)))
+        (or (file-request request root-path options) (handler request))))
      ([request respond raise]
-      (if-let [response (file-request request root-path options)]
-        (respond response)
-        (handler request respond raise))))))
+      (if (:prefer-handler? options)
+        (let [handler-response (promise)]
```

Comment:
You can't use a promise here. The three-argument form of handlers needs to be asynchronous, and promises block. You need to put the response logic in the callback.